kroxylicious/src/main/templates/Kproxy/PerMessageInvoker.ftl

```diff
@@ -0,0 +1,116 @@
+    </#list>
+    public PerMessageInvoker(Object filter){
+        if(filter instanceof KrpcFilter){
+            this.inner = (KrpcFilter) filter;
+        } else {
+            this.inner = null;
+        }
+        Set<ApiKeys> supportedRequestKeys = new HashSet<>();
+        Set<ApiKeys> supportedResponseKeys = new HashSet<>();
+        <#list messageSpecs as messageSpec>
+        ${messageSpec.name}Filter = filter instanceof ${messageSpec.name}Filter ? (${messageSpec.name}Filter) filter : null;
+        <#if messageSpec.type?lower_case == 'request'>
+        if(filter instanceof ${messageSpec.name}Filter) {
+          supportedRequestKeys.add(ApiKeys.${retrieveApiKey(messageSpec)});
+        }
+        </#if>
+        <#if messageSpec.type?lower_case == 'response'>
+        if(filter instanceof ${messageSpec.name}Filter) {
+          supportedResponseKeys.add(ApiKeys.${retrieveApiKey(messageSpec)});
+        }
+        </#if>
+        </#list>
+        requestKeys = supportedRequestKeys.isEmpty() ? EnumSet.noneOf(ApiKeys.class): EnumSet.copyOf(supportedRequestKeys);
+        responseKeys = supportedResponseKeys.isEmpty() ? EnumSet.noneOf(ApiKeys.class): EnumSet.copyOf(supportedResponseKeys);
+    }
+
+    public void onRequest(DecodedRequestFrame<?> decodedFrame, KrpcFilterContext filterContext) {
+        switch (decodedFrame.apiKey()) {
+        <#list messageSpecs as messageSpec>
+        <#if messageSpec.type?lower_case == 'request'>
+            case ${retrieveApiKey(messageSpec)}:
+                ${messageSpec.name}Filter.on${messageSpec.name}((${messageSpec.name}Data) decodedFrame.body(), filterContext);
+                break;
+        </#if>
+        </#list>
+            default:
+                throw new IllegalStateException("Unsupported RPC " + decodedFrame.apiKey());
+        }
+    }
+
+    public void onResponse(DecodedResponseFrame<?> decodedFrame, KrpcFilterContext filterContext) {
+        switch (decodedFrame.apiKey()) {
+        <#list messageSpecs as messageSpec>
+        <#if messageSpec.type?lower_case == 'response'>
+            case ${retrieveApiKey(messageSpec)}:
+                ${messageSpec.name}Filter.on${messageSpec.name}((${messageSpec.name}Data) decodedFrame.body(), filterContext);
+                break;
+        </#if>
+        </#list>
+            default:
+                throw new IllegalStateException("Unsupported RPC " + decodedFrame.apiKey());
+        }
+    }
+
+    public boolean shouldDeserializeRequest(ApiKeys apiKey, short apiVersion) {
+        return requestKeys.contains(apiKey) && (inner == null || inner.shouldDeserializeRequest(apiKey, apiVersion));
+    }
+
+    public boolean shouldDeserializeResponse(ApiKeys apiKey, short apiVersion) {
+        return responseKeys.contains(apiKey) && (inner == null || inner.shouldDeserializeResponse(apiKey, apiVersion));
+    }
+
+    public static boolean implementsAnyMessageInterface(Object obj){
```

Comment:
If we're keeping `KrpcFilter` as the base interface for the per-request and response interfaces then we should declare the parameter as a `KrpcFilter`: It's then the caller's job (specifically the `FilterChainFactory`) to do a downcast. It is anyway in a better position to provide a sensible exception message (relating it back to a specific config).